### ISSUE TITLE
type guard does not narrow intersection over union

### ISSUE THREAD

**Issue by siegebell** [Microsoft/TypeScript#12180]
**TypeScript Version:**  2.0.8
**Code**

```ts
interface A {tag: "a", a};
interface B {tag: "b", b};
interface C {c};
type ABC1 = (A & C) | (B & C);
let y : ABC1 = {tag:"a", a: "foo", c: "bar"}
if(y.tag === "a")
  y.a // this is accepted
type ABC2 = (A | B) & C;
let x : ABC2 = {tag:"a", a: "foo", c: "bar"}
if(x.tag === "a")
  x.a // but this is currently an error :(
```

**Expected behavior:**
After checking `x.tag==="a"`, the type of `x` should be narrowed to `A & C`
**Actual behavior:** 
There is no narrowing, so `x.a` results in a type error.

A workaround is given by `ABC1`, but in practice, this results in very long and redundant type definitions.

**Comment by siegebell** [comment]
Hmm. It seems that `ABC1` and `ABC2` are not even considered equivalent types:
```ts
x = y; // okay: y:ABC1 is a subtype of x:ABC2
y = x; // error
```

Is this intended behavior? I.e. is there a case where `x` and `y` might not be compatible?

**Comment by RyanCavanaugh** [comment]
This has been fixed by #11717

**Comment by siegebell** [comment]
@RyanCavanaugh phew -- Thanks!
